### PR TITLE
feat: .fromCell() and .fromSlice() functions

### LINKED ISSUE #260: Add `Struct.fromCell` and `Struct.fromSlice` extension functions
See: https://github.com/tact-lang/tact/pull/418

### PR DESCRIPTION
**Pull request by novusnota**
Closes #260

### CODE REVIEW

File: pages/book/structs-and-messages.mdx [hunk 1]
```diff
@@ -198,9 +198,43 @@ fun conversionFun() {
...
+  [`Message.toCell(){:tact}`](/ref/core-cells#messagetocell)
 
 </Callout>
 
+### Obtain from a `Cell` or `Slice`, `.fromCell()` and `.fromSlice()` [#fromcellslice]
+
+Instead of manually parsing a [`Cell{:tact}`][p] or [`Slice{:tact}`][p] via the series of relevant `.loadSomething(){:tact}` function calls, one can use `.fromCell(){:tact}` and `.fromSlice(){:tact}` [extension functions](/book/functions#extension-function) for converting the provided [`Cell{:tact}`][p] or [`Slice{:tact}`][p] into the needed [Struct](#structs) or [Message](#messages).
+
+Those extension functions only attempt to parse a [`Cell{:tact}`][p] or [`Slice{:tact}`][p] according to the structure of your [Struct](#structs) or [Message](#messages). In case layouts don't match, various exceptions may be thrown — make sure to wrap your code in [`try...catch{:tact}`](/book/statements#try-catch) blocks to prevent unexpected results.
+
```
Comment (anton-trunov) [inline pages/book/structs-and-messages.mdx:211]: would be nice to add the following laws hold:

- for any `struct` of type `Struct` (which is a struct or a message):

```
Struct.fromCell(struct.toCell()) == struct
```
and

- for any `cell` and type `Struct` (which is a struct or a message), if `Struct.fromCell(cell)` does not throw, then

```
Struct.fromCell(cell).toCell() == cell
```

and the same goes for `toSlice` / `fromSlice`.

File: pages/ref/core-cells.mdx [hunk 2]
```diff
@@ -462,6 +462,27 @@ let s: Slice = beginCell().storeBool(true).asSlice();
 let fizz: Bool = s.loadBool(); // true
 ```
 
+## Slice.loadBool
```
Comment (anton-trunov) [inline pages/ref/core-cells.mdx:465]: but we just merged https://github.com/tact-lang/tact-docs/pull/270

File: pages/ref/core-cells.mdx [hunk 3]
```diff
@@ -733,6 +754,76 @@ fun coinCell(): Cell {
...
+}
+```
+
+## Struct.fromSlice
+
+```tact
+extends fun fromSlice(self: Struct, cell: Slice): Struct;
+```
+
+Extension function for any [Struct][struct].
```
Comment (anton-trunov) [inline pages/ref/core-cells.mdx:798]:
```suggestion
Extension function for any structure type [Struct][struct].
```

File: pages/ref/core-cells.mdx [hunk 4]
```diff
@@ -733,6 +754,76 @@ fun coinCell(): Cell {
 }
 ```
 
+## Struct.fromCell
+
+```tact
+extends fun fromCell(self: Struct, cell: Cell): Struct;
+```
+
+Extension function for any [Struct][struct].
```
Comment (anton-trunov) [inline pages/ref/core-cells.mdx:763]:
```suggestion
Extension function for any structure type [Struct][struct].
```

File: pages/ref/core-cells.mdx [hunk 5]
```diff
@@ -759,6 +850,75 @@ fun coinCell(): Cell {
 }
 ```
 
+## Message.fromCell
+
+```tact
+extends fun fromCell(self: Message, cell: Cell): Message;
+```
+
+Extension function for any [Message][message].
```
Comment (anton-trunov) [inline pages/ref/core-cells.mdx:859]:
```suggestion
Extension function for any message type [Message][message].
```

File: pages/ref/core-cells.mdx [hunk 6]
```diff
@@ -759,6 +850,75 @@ fun coinCell(): Cell {
...
+```
+
+## Message.fromSlice
+
+
+```tact
+extends fun fromSlice(self: Message, cell: Slice): Message;
+```
+
+Extension function for any [Message][message].
```
Comment (anton-trunov) [inline pages/ref/core-cells.mdx:894]:
```suggestion
Extension function for any message type [Message][message].
```

**Review by anton-trunov — CHANGES_REQUESTED** [review]

File: pages/ref/core-cells.mdx (hunk 2 above)
Comment (novusnota) [reply pages/ref/core-cells.mdx:465]: Hmm, seems like the older one was brought over. Tried to fix with rebasing from the main, still there :/

**Comment by novusnota** [comment]
Seems like I broke this and #272 pull requests. Will close those and re-create again, with suggestions applied.

File: pages/book/structs-and-messages.mdx (hunk 1 above)
Comment (novusnota) [reply pages/book/structs-and-messages.mdx:211]: @anton-trunov Those are nice, but:

* `toSlice(){}` doesn't exist. One can do `Struct.toCell().asSlice()` though
* Equality operator `==` cannot compare Structs and Messages directly. One would need to compare them as Cells (at the moment)

Comment (anton-trunov) [reply pages/book/structs-and-messages.mdx:211]: Good point about `toSlice()` -- we should probably add it, for consistency reasons, e.g. `StringBuilder` has it.

The equality `==` operator is meant here in the mathematical sense, it's an instance of equational reasoning which is used commonly in functional programming and `=` seemed to be a bit inappropriate given its imperative meaning. We might want to use `===`, add an explanation here, or use LaTeX and possible quantifiers: `∀ s: Struct, Struct.fromCell(s.toCell()) = s`

**Pull request closed without merging**